compiler/src/dotty/tools/dotc/typer/Inferencing.scala

```diff
@@ -411,10 +409,16 @@ object Inferencing {
     Stats.record("maximizeType")
     val vs = variances(tp)
     val patternBindings = new mutable.ListBuffer[(Symbol, TypeParamRef)]
+    val gadtBounds = ctx.gadt.symbols.map(ctx.gadt.bounds(_).nn)
     vs foreachBinding { (tvar, v) =>
       if !tvar.isInstantiated then
-        if (v == 1) tvar.instantiate(fromBelow = false)
-        else if (v == -1) tvar.instantiate(fromBelow = true)
+        // if the tvar is covariant/contravariant (v == 1/-1, respectively) in the input type tp
+        // then check the tvar doesn't occur in the opposite GADT bound (lower/upper) within any of the GADT bounds
+        // if it doesn't occur then it's safe to instantiate the tvar
+        // Eg neg/i14983 the C in Node[+C] is in the GADT lower bound X >: List[C] so maximising to Node[Any] is unsound
+        // Eg pos/precise-pattern-type the T in Tree[-T] is in no GADT upper bound so can maximise to Tree[Type]
+        val safeToInstantiate = v != 0 && gadtBounds.forall(tb => !tvar.occursIn(if v == 1 then tb.lo else tb.hi))
```

Comment:
This doesn't entirely feel correct, we're not taking into account the variance at which `tvar` is occuring at in `tb`. I would disallow instantiating if `tvar` occurs at all in `tb`.

Reply:
That's correct.  And I was hoping to find a legitimate case in either bootstrapping or in the community build which would help me figure out what variance flipping I need to be doing.  But I didn't encounter any.  Do you have in mind an example where this would be negatively impacted.

I proposed this version because it fixes a soundness issue even if it might have this incompleteness issue (correct me if I'm misdiagnosing).

Reply:
This indeed results in soundness problems. Consider the following example, whose main difference with #14983 is that the value encapsulated in `Node` has a contravariant type argument.

```scala
case class Showing[-C](show: C => String)

sealed trait Tree[+A]
final case class Leaf[+B](b: B)       extends Tree[B]
final case class Node[-C](l: Showing[C]) extends Tree[Showing[C]]

object Test:
  def meth[X](tree: Tree[X]): X = tree match
    case Leaf(v) => v
    case Node(x) =>
      // GADT bounds: X >: Showing[C]
      // maximise Node[C] to Node[Nothing].
      //   Since C is contravariant and in GADT lower bound,
      //   instantiation is currently assumed to be safe.
      // After that: X >: Showing[Nothing]
      Showing[String](_ + " boom!")

  def main(args: Array[String]): Unit =
    val tree = Node(Showing[Int](_.toString))
    val res = meth(tree)
    println(res.show(42))  // runtime crash: ClassCastException
```

The problem here is that the type variable `C` are contravariant in both places. So maximising `Showing[C]` also results in maximising the GADT lower bound of `X`, which makes the constraint unsound.

Reply:
Maybe we should disable the instantiation as long as the type variable appears in GADT bounds as suggested by Alex? Or shall we query the type variable variances in GADT bounds using `variances`, and only instantiate the tvar when its variance conforms with that in GADT bounds?

Reply:
If we want to take variance into account, we will need an argument why that's sound.

The way I understand things, the type variable is supposed to correspond to a type member in DOT terms: type variables in PatternTypeConstrainer will only come from type arguments to `unapply` functions, and in the pattern type they will always appear as type arguments to some class, unless we're dealing with an `unapply` with a very strange signature. If we do instantiate the type variable, we turn it into an abstract type, which makes it clearly correspond to the type member. In _certain_ situations it's OK to approximate the type variable to avoid unnecessary problems with things like implicit search, but if the type variable appears in the constraint, approximating the type variable could be unsound.